File: 09.py
#! /usr/bin/env python

from time import time
import psutil

_id = 0


class Marble:
    __slots__ = ('val', 'prev', 'next')

    def __init__(self, val, prev=None, next_=None):
        self.val = val
        self.prev = prev or self
        self.next = next_ or self


def get_high_score(nb_players, last_marble):
    marble = Marble(0, None, None)
    scores = {}

    p = 0

    for v in range(1, last_marble+1):
        if v % 1000000 == 0:
            print('%8d | %5.1f%%' % (v, v / last_marble * 100))

        if v % 23 == 0:
            for _ in range(7):
                marble = marble.prev

            scores[p] = scores.get(p, 0) + v + marble.val

            prev = marble.prev
            next_ = marble.next

            prev.next = next_
            next_.prev = prev
            marble = next_
        else:
            marble = marble.next.next
            prev = marble.prev

            new_m = Marble(v, prev, marble)

            prev.next = new_m
            marble.prev = new_m

            marble = new_m

        p += 1

        if p == nb_players:
            p = 0

    print('Memory: %s' % get_human_readable_size(psutil.Process().memory_info().rss))

    return max(scores.values())


def get_human_readable_size(size):
    for unit in ('B', 'KiB', 'MiB', 'GiB', 'TiB'):
        if size < 1024:
            return '%.2f %s' % (size, unit)
        elif unit == 'TiB':
            return '%d %s' % (round(size), unit)

        size /= 1024


def main():
    print("Part 1: %d" % get_high_score(458, 72019))
    print("Part 2: %d" % get_high_score(458, 7201900))


if __name__ == '__main__':
    ts = time()
    main()
    ts = time() - ts
    print('Done in %.3fms' % (ts * 1000))

File: 09.c
#include <stdio.h>
#include <stdlib.h>

struct marble {
    long val;
    struct marble *next;
    struct marble *prev;
};

typedef struct marble marble_s;

long get_high_score(int nb_players, int num_marbles) {
    int c = 0;
    marble_s *marbles = malloc(sizeof(marble_s) * num_marbles);

    marble_s *marble = &marbles[c++];
    marble->val = 0;
    marble->next = marble;
    marble->prev = marble;

    long *scores = calloc(nb_players, sizeof(long));

    int p = 0;

    for (long v=1; v<=num_marbles; v++) {
        if (v % 23 == 0) {
            for (int i=0; i<7; i++) {
                marble = marble->prev;
            }

            scores[p] += (v + marble->val);

            marble_s *prev = marble->prev;
            marble_s *next = marble->next;

            prev->next = next;
            next->prev = prev;
            marble = next;
        } else {
            marble = marble->next->next;

            marble_s *prev = marble->prev;

            marble_s *new_m = &marbles[c++];
            new_m->val = v;
            new_m->prev = prev;
            new_m->next = marble;

            prev->next = new_m;
            marble->prev = new_m;

            marble = new_m;
        }

        p += 1;

        if (p == nb_players) {
            p = 0;
        }
    }

    long hs = 0;
    for (int i=0; i<nb_players; i++) {
        if (scores[i] > hs) {
            hs = scores[i];
        }
    }

    free(marbles);
    free(scores);

    return hs;
}

int main(void) {
    printf("Part 1: %ld\n", get_high_score(458, 72019));
    printf("Part 2: %ld\n", get_high_score(458, 7201900));
}
